### PR TITLE
Implementation of two-factor authentication and IVAs

### PR DESCRIPTION
**Pull request by dontseyit**
This PR updates the test bed to align with the new authentication process. For further details, please read [the technical specifications](https://github.com/ghga-de/epic-docs/blob/main/46-sugar-ant/technical_specification.md) of the original implementation. 

External dependencies are also upgraded therefore some scenarios are refactored accordingly to avoid fixture mismanagement issues. Please see this [PR](https://github.com/ghga-de/archive-test-bed/pull/77)

To ensure expected behavior and integration of two-factor authentication and IVA management, three key requirements must be incorporated into the test bed:

1. **Session Management**
The authentication fixture in the test bed has been revised to handle session management rather than token-based authentication. Access to an internal service still uses an internal token; this aspect of the test bed remains mostly unchanged, except that now a session is used to generate internal token. The method previously used to manage request headers across different scenarios has been separated into distinct actions. Also session data is now stored in state storage, allowing to access user session at any step once the user is logged in. This also facilitates adjustments for scenarios involving multiple concurrent users or in cases of unhappy test cases.

2. **Two-Factor Authentication** 
Users are required to complete two-factor authentication upon their login and registration. The user login mechanism and corresponding scenarios have been updated to accommodate this requirement.

3. **IVA Management**
Users are required to submit Independent Verification Addresses (IVAs) to confirm their identity. Tests for the access requests and work package creation are updated to incorporate IVA management. 

Below **document** shows the implementation details of the new scenarios and state management throughout the steps. Only the details relevant to 2FA and IVA implementation are shown.
Up to date as of commit [bfbda6c](https://github.com/ghga-de/archive-test-bed/pull/78/commits/bfbda6c1f2f949afee1c938da90a3a9df7189ead).
image

### CODE REVIEW

File: features/33_work_packages.feature [hunk 1]
```diff
@@ -1,29 +1,29 @@
 @download @wps
-Feature: 32 Work Packages
+Feature: 33 Work Packages
   As a user, I can create a work package for downloading a file
   and a download token corresponding ot that work package.
 
   Background:
     Given we have the state "John Doe is allowed to download the test dataset"
     And I am registered as "Dr. John Doe"
     And I am logged in as "Dr. John Doe"
```
Comment (Cito) [inline features/33_work_packages.feature:9]:
```suggestion
    And I am logged in as "Dr. John Doe"
    And I am registered as "Dr. John Doe"
```

to show the proper order

File: steps/test_30_user_registration.py [hunk 2]
```diff
@@ -77,7 +92,7 @@ def user_gets_id(full_name: str, fixtures: JointFixture, response: Response):
     user_id = user["id"]
     assert user_id and "-" in user_id and len(user_id) > 6 and "@" not in user_id
     assert user["name"] == name
-    assert user["title"] == title
+    assert user["title"] == title  # FIXME: title is None, but should be "Dr."
```
Comment (Cito) [inline steps/test_30_user_registration.py:95]:
```suggestion
    assert user["title"] == title  # FIXME: title is None, but should be "Dr."
```

Is this comment still relevant?

File: steps/test_31_user_verification.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    iva = fixtures.state.get_state("iva")
+    session = fixtures.auth.get_session(name=full_name, state_store=fixtures.state)
+    assert session, "User session not found"
+    headers = fixtures.auth.headers(session=session) if session else {}
+    assert iva["id"], "IVA ID not found"
+    return fixtures.iva.create_verification(iva["id"], headers=headers)
+
+
+@then(parse('"{full_name}" sends the verification code to the IVA'))
+def sent_verification_code(verification_code: str, fixtures: JointFixture):
```
Comment (Cito) [inline steps/test_31_user_verification.py:109]:
```suggestion
def send_verification_code(verification_code: str, fixtures: JointFixture):
```

File: steps/test_31_user_verification.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    headers = fixtures.auth.headers(session=session) if session else {}
+    results = fixtures.iva.retrieve(session.user_id, headers)
+    ivas_to_delete = [iva["id"] for iva in results]
+    fixtures.iva.delete(ivas_to_delete, session.user_id, headers)
+
+
+@when(parse('"{full_name}" adds "{prop}" as an IVA'), target_fixture="iva")
+def user_adds_an_iva(full_name: str, prop: str, fixtures: JointFixture) -> IVA:
+    """Add an IVA to the user account"""
+    iva_value = USER_IVAS.get(prop, None)
```
Comment (Cito) [inline steps/test_31_user_verification.py:57]:
```suggestion
    iva_value = USER_IVAS.get(prop)
```

None is the default

File: steps/test_40_delete_datasets.py [hunk 5]
```diff
@@ -155,10 +154,15 @@ def check_access_grants_in_claims_repository(config: Config, mongo: MongoFixture
...
-# The following step is re-used from the work package creation test
-when("the list of datasets is queried", target_fixture="response")(
-    query_datasets_with_wps
-)
+@when(parse('"{full_name}" lists the datasets'), target_fixture="response")
+def query_datasets_with_wps(fixtures: JointFixture, full_name: str):
+    # This step is the copy of the one in the work package creation test
+    session = fixtures.auth.get_session(name=full_name, state_store=fixtures.state)
+    assert session
+    headers = fixtures.auth.headers(session=session) if session else {}
```
Comment (Cito) [inline steps/test_40_delete_datasets.py:162]: Maybe make headers accept an empty session and return an empty dict in that case to simplify these constructs.

**Review by Cito — CHANGES_REQUESTED** [review]
Very nice. Good to see that everything is working.

Made some suggestions below.

File: fixtures/auth.py [hunk 6]
```diff
@@ -153,42 +179,204 @@ def internal_access_token_from_name(
...
+        if self.use_api_gateway or self.use_auth_adapter:
+            headers = self.headers_for_session(session=session)
             if not self.use_api_gateway:
-                token = self.internal_access_token_from_external_access_token(
-                    token, for_registration=not user_id
-                )
+                internal_token = self.internal_access_token_from_session(headers)
+                headers = {"Authorization": f"Bearer {internal_token}"}
+        else:
+            # Is there any scenario where we don't have an Auth Adapter?
```
[the remark this answers is not in the document]
Comment (dontseyit) [reply fixtures/auth.py]: With the current state of the test bed, without an auth adapter, some tests fail. So, if we really need to include this scenario, we should implement it properly (e.g., with 2FA operations, session states); in this case, we should address it in a separate PR.

[the remark this answers is not in the document]
Comment (Cito) [reply fixtures/auth.py]: Yes, the test bed would not only have to emulate the token exchange, but also the TOTP functionality of the auth adapter, or skip these tests in that scenario.

Again, my feeling is that we should keep it simple and not support this scenario any longer, i.e. assume that we either test via API gateway (remote tests) or have the auth adapter directly accessible (local tests).

[the remark this answers is not in the document]
Comment (dontseyit) [reply fixtures/auth.py]: I agree. I've removed the `use_auth_adapter` config and related code.

**Review by Cito — APPROVED** [review]
Thanks, looks good.

**Pull request merged**